Improve suggestions for `NonZeroT` <- `T` coercion error

WaffleLapkin opened the pull request:
Currently, when encountering a type mismatch error with `NonZeroT` and `T` (for example `NonZeroU8` and `u8`) we errorneusly suggest wrapping expression in `NonZeroT`:
```text
error[E0308]: mismatched types
 --> ./t.rs:7:35
  |
7 |     let _: std::num::NonZeroU64 = 1;
  |            --------------------   ^ expected struct `NonZeroU64`, found integer
  |            |
  |            expected due to this
  |
help: try wrapping the expression in `std::num::NonZeroU64`
  |
7 |     let _: std::num::NonZeroU64 = std::num::NonZeroU64(1);
  |                                   +++++++++++++++++++++ +
```

I've removed this suggestion and added suggestions to call `new` (for `Option<NonZeroT>` <- `T` case) or `new` and `unwrap` (for `NonZeroT` <- `T` case):

```text
error[E0308]: mismatched types
 --> ./t.rs:7:35
  |
7 |     let _: std::num::NonZeroU64 = 1;
  |            --------------------   ^ expected struct `NonZeroU64`, found integer
  |            |
  |            expected due to this
  |
help: Consider calling `NonZeroU64::new`
  |
7 |     let _: std::num::NonZeroU64 = NonZeroU64::new(1).unwrap();
  |                                   ++++++++++++++++ ++++++++++

error[E0308]: mismatched types
 --> ./t.rs:8:43
  |
8 |     let _: Option<std::num::NonZeroU64> = 1;
  |            ----------------------------   ^ expected enum `Option`, found integer
  |            |
  |            expected due to this
  |
  = note: expected enum `Option<NonZeroU64>`
             found type `{integer}`
help: Consider calling `NonZeroU64::new`
  |
8 |     let _: Option<std::num::NonZeroU64> = NonZeroU64::new(1);
  |                                           ++++++++++++++++ +
```

r? @compiler-errors

WaffleLapkin committed:
Do not suggest "wrapping the expression in `std::num::NonZeroU64`"

WaffleLapkin committed:
Suggest a fix for `NonZero*` <- `*` coercion error

compiler-errors commented on compiler/rustc_typeck/src/check/demand.rs:
```diff
@@ -347,6 +349,13 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
             }
 
+            // Avoid suggesting wrapping in `NonZeroU64` and alike
+            if self.tcx.layout_scalar_valid_range(expected_adt.did())
```
Is it possible to get `the_adt.fields[0]` and check its privacy (using [this method](https://doc.rust-lang.org/nightly/nightly-rustc/rustc_middle/ty/enum.Visibility.html#method.is_accessible_from) maybe, and the def id of the fcx body) instead? Or maybe best to put this privacy logic in `compatible_variants` below.

It might also help us not suggest in case like:

```
mod inner {
  pub struct Wrapper<T>(T);
}

fn needs_wrapper(t: Wrapper<i32>) {}

fn main() {
  needs_wrapper(0);
}
```

WaffleLapkin replied:
I was thinking of doing this, but couldn't figure how to check for field accessibility 😅 

I'll try to use [`is_accessible_from`](https://doc.rust-lang.org/nightly/nightly-rustc/rustc_middle/ty/enum.Visibility.html#method.is_accessible_from) for this now

WaffleLapkin replied:
Done ;)

compiler-errors reviewed (commented):
Thanks @WaffleLapkin for an interesting and useful diagnostics improvement as always. This looks good as is, but I had an idea of generalizing both parts of the PR. Let me know your thoughts.

compiler-errors commented on compiler/rustc_typeck/src/check/demand.rs:
```diff
@@ -410,6 +419,59 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
         }
     }
 
+    fn suggest_non_zero_new_unwrap(
```
I guess in the case of `NonZero*` the `new` signature is not `{expr_ty} -> {expected}` but `{expr_ty} -> Option<{expected}>` ... maybe the non-zero suggestion can stay as is.

compiler-errors replied:
Yeah, sounds good.

WaffleLapkin committed:
check accessibility before suggesting wrapping expressions

compiler-errors commented on compiler/rustc_typeck/src/check/demand.rs:
```diff
@@ -355,6 +356,16 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
                 .filter_map(|variant| {
                     let sole_field = &variant.fields[0];
+
+                    if !sole_field.did.is_local()
+                        && !sole_field.vis.is_accessible_from(
+                            self.tcx.parent_module(expr.hir_id).to_def_id(),
+                            self.tcx,
+                        )
```
```suggestion
                    if !sole_field.vis.is_accessible_from(
                            self.body_id.owner.to_def_id(),
                            self.tcx,
                        )
```

compiler-errors replied:
No need to restrict to only foreign def id, and we can get a more accurate def id from `self.body_id.owner.to_def_id()` i think.

WaffleLapkin replied:
Restriction to foreign def ids is deliberate, see the test: https://github.com/rust-lang/rust/blob/da2752e00f0139fb13282b2bd97aa0f8c665aee9/src/test/ui/mismatched_types/wrap-suggestion-privacy.rs#L10-L11

compiler-errors replied:
Specifically " its privacy can be easily changed " as the justification would be nice to note

WaffleLapkin replied:
Not sure how to present it in a good way, but I've tried 5bd88dfa8a0b5d5bcdd84d21e2f465719ce8c328

compiler-errors replied:
haha I meant a comment in the source code, but this works too!

WaffleLapkin replied:
lmao

it's bed time

compiler-errors commented on compiler/rustc_typeck/src/check/demand.rs:
```diff
@@ -410,6 +421,59 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
+            // In case Option<NonZero*> is wanted, but * is provided, suggest calling new
+            ty::Adt(adt, substs) if tcx.is_diagnostic_item(sym::Option, adt.did()) => {
+                // Unwrap option
+                let Some(fst) = substs.first() else { return };
+                let ty::Adt(adt, _) = fst.expect_ty().kind() else { return };
```
```suggestion
                let ty::Adt(adt, _) = fst.type_at(0).kind() else { return };
```

compiler-errors commented on compiler/rustc_typeck/src/check/demand.rs:
```diff
@@ -410,6 +421,59 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
+        let path = self.tcx.def_path_str(adt.non_enum_variant().def_id);
+
+        err.multipart_suggestion(
+            format!("consider calling `{s}::new`"),
+            vec![
+                (expr.span.shrink_to_lo(), format!("{path}::new(")),
```
What's the difference between `{s}` and `{path}`? Can we just use one of them?

WaffleLapkin replied:
`path` can be `std::num::NonZeroU64`, while `s` is just `NonZeroU64`

WaffleLapkin commented on compiler/rustc_typeck/src/check/demand.rs:
```diff
@@ -410,6 +421,59 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
+
+        let Some((s, _)) = map
+            .iter()
+            .find(|&&(s, _)| self.tcx.is_diagnostic_item(s, adt.did()))
+            .filter(|&&(_, t)| { self.can_coerce(expr_ty, t) })
```
mh?

compiler-errors replied:
as in, `.find(|&&(s, t)| self.tcx.is_diagnostic_item(s, adt.did()) && self.can_coerce(expr_ty, t))`

WaffleLapkin replied:
This would continue searching even after the matching diagnostic item was reached... though ig it doesn't matter "performance wise", given a list of 10 elements

compiler-errors replied:
I see, either way up to you then, but I don't think we need to optimize diagnostics code.

WaffleLapkin committed:
Apply suggestions from the review

- Use `expr.hir_id.owner` instead of `self.tcx.parent_module(expr.hir_id)`
- Use `.type_at()` instead of `.first()` + `.expect_ty()`
- Use single `.find()` with `&&` condition

Co-authored-by: Michael Goulet <michael@errs.io>

WaffleLapkin committed:
Add a note about privacy to wrapping suggestion

compiler-errors reviewed (approved):
This looks good for now. It can be tweaked later if anyone complains about it.

compiler-errors commented:
r=me once CI turns green

@bors delegate+

bors commented:
:v: @WaffleLapkin can now approve this pull request
<!-- homu: {"type":"Delegated","delegator":"compiler-errors","delegate":"WaffleLapkin"} -->

WaffleLapkin commented:
@bors r=compiler-errors

bors commented:
:pushpin: Commit 5bd88dfa8a0b5d5bcdd84d21e2f465719ce8c328 has been approved by `compiler-errors`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=compiler-errors 5bd88dfa8a0b5d5bcdd84d21e2f465719ce8c328 -->
<!-- homu: {"type":"Approved","sha":"5bd88dfa8a0b5d5bcdd84d21e2f465719ce8c328","approver":"compiler-errors","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->
